Don't allow rsvp-ing a workshop that is in the past #1484

simbabque opened the pull request:
The backend did not do validation on the workshop's time and date when
a user sent an rsvp. There is validation in the view when the workshop
page is rendered, so the button to attend does not show up on past
workshops. However if the page was old and the button was there, a user
could still sign up and get an invitation created. The workshop was then
displayed as in the past.

For virtual workshops it is possible to attend anyway, so a confirmation
email would be sent. This change stops the system from ever creating
invitations for past workshops.

I think the new test file should also contain tests for the other validation criteria in the `workshops_controller#rsvp action`. I decided not to include them here to keep the PR to the scope of the issue.

Closes #1484

simbabque commented:
I'm not sure how to address the two issues that CodeFactor and codeclimate have found. This is my first Ruby code, and I'm not familiar with the tools.

I believe I could have make the `rsvp` method shorter by switching existing code to post-fix conditions, but that feels like it would reduce readability, thus defeating what I think the point of these tools is. I don't know what to do about too many expectations. I've always struggled when I was told my expectations are too high. :wink:

despo commented on app/controllers/workshops_controller.rb:
```diff
@@ -11,6 +11,10 @@ def show
     redirect_to :back, notice: t('workshops.registration_not_open') unless @workshop.invitable_yet?
 
+    unless @workshop.date_and_time.future?
+      return redirect_to :back, notice: t('messages.already_taken_place')
```
```suggestion
      return redirect_to :back, notice: t('messages.already_taken_place')  unless @workshop.date_and_time.future?
```

despo replied:
@simbabque Maybe a better way around this that is also easier to read (and should resolve the codeclimate issue) is to extend the workshop model to perform the check about the workshop being available for rsvp, and calling that through here. I think adding a more specific message is a nice to have, but not essential as I _think_ the refreshed page might already have some text indicating that the event has already taken place.


e.g.

```ruby
def available_for_rsvp?
   invitable_yet? && date_and_time.future?
end

despo commented on spec/features/rsvping_workshop.rb:
```diff
@@ -0,0 +1,37 @@
+      expect(page).not_to have_button('Attend as a coach')
+    end
+  end
+end
```
We already have a spec testing this functionality https://github.com/codebar/planner/blob/master/spec/features/managing_workshop_attendance_spec.rb so I we only need the last scenario. It should probably be added to the [shared examples](https://github.com/codebar/planner/blob/master/spec/support/shared_examples/behaves_like_managing_workshop_attendance.rb) file.

despo reviewed (changes requested):
Thank you for your PR. I left some feedback about how the codefactor and codeclimate issues can be addressed and some additional comments about the specs.

simbabque commented:
Thank you @despo. I've moved the check into the model and moved the test into existing shared examples file. While there I noticed that another context that checks for the attendance buttons with `have_link` could never pass, so I included a fix for that.

despo commented on spec/support/shared_examples/behaves_like_managing_workshop_attendance.rb:
```diff
@@ -163,10 +163,21 @@
-            expect(page).not_to have_link('Attend as a student')
-            expect(page).not_to have_link('Attend as a coach')
+            expect(page).not_to have_button('Attend as a student')
+            expect(page).not_to have_button('Attend as a coach')
```
Good catch! Thank you.

despo reviewed (approved):
@simbabque this looks great. Thank you so much! 💯
